File: 백준/Gold/10026. 적록색약/적록색약.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        String[] spaces = new String[n];
        for (int i = 0; i < n; i++) {
            spaces[i] = br.readLine();
        }
        boolean[][] visited = new boolean[n][n];
        boolean[][] combineVisited = new boolean[n][n];
        int[][] MOVES = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        int normal = 0, combined = 0;
        Deque<int[]> stack = new LinkedList<>();

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (!visited[i][j]) {
                    stack.add(new int[]{i, j});
                    visited[i][j] = true;
                    normal += 1;
                    while (!stack.isEmpty()) {
                        int[] xy = stack.pop();
                        for (int[] move : MOVES) {
                            int ny = xy[0] + move[0], nx = xy[1] + move[1];
                            if ((0 <= ny && ny < n) && (0 <= nx && nx < n)
                                    && !visited[ny][nx] && spaces[ny].charAt(nx) == spaces[xy[0]].charAt(xy[1])) {
                                stack.push(new int[]{ny, nx});
                                visited[ny][nx] = true;
                            }
                        }
                    }
                }

                if (!combineVisited[i][j]) {
                    stack.add(new int[]{i, j});
                    combineVisited[i][j] = true;
                    combined += 1;
                    while (!stack.isEmpty()) {
                        int[] xy = stack.pop();
                        for (int[] move : MOVES) {
                            int ny = xy[0] + move[0], nx = xy[1] + move[1];
                            if ((0 <= ny && ny < n) && (0 <= nx && nx < n)
                                    && !combineVisited[ny][nx] && (spaces[ny].charAt(nx) == spaces[xy[0]].charAt(xy[1])
                                    || (spaces[ny].charAt(nx) != 'B' && spaces[xy[0]].charAt(xy[1]) != 'B'))) {
                                stack.push(new int[]{ny, nx});
                                combineVisited[ny][nx] = true;
                            }
                        }
                    }
                }
            }
        }

        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        bw.write(normal + " " + combined);
        bw.flush();
    }
}

File: 백준/Gold/10026. 적록색약/적록색약.py
from sys import stdin

n = int(stdin.readline())
spaces = [' '*(n+2)] + [' ' + stdin.readline() + ' ' for _ in range(n)] + [' '*(n+2)]
visited = [[True]*(n+2)] + [[True] + [False]*n + [True] for _ in range(n)] + [[True]*(n+2)]
combine_visited = [[True]*(n+2)] + [[True] + [False]*n + [True] for _ in range(n)] + [[True]*(n+2)]
MOVES = [(-1, 0), (1, 0), (0, -1), (0, 1)]

normal, combined = 0, 0
stack = []

for i in range(1, n+1):
    for j in range(1, n+1):
        if not visited[i][j]:
            stack.append((i, j))
            visited[i][j] = True
            normal += 1
            while stack:
                y, x = stack.pop()
                for dy, dx in MOVES:
                    ny, nx = y + dy, x + dx
                    if not visited[ny][nx] and spaces[ny][nx] == spaces[y][x]:
                        stack.append((ny, nx))
                        visited[ny][nx] = True

        if not combine_visited[i][j]:
            stack.append((i, j))
            combine_visited[i][j] = True
            combined += 1
            while stack:
                y, x = stack.pop()
                for dy, dx in MOVES:
                    ny, nx = y + dy, x + dx
                    if not combine_visited[ny][nx] and (spaces[ny][nx] == spaces[y][x] or 'B' not in (spaces[ny][nx], spaces[y][x])):
                        stack.append((ny, nx))        
                        combine_visited[ny][nx] = True


print(normal, combined)